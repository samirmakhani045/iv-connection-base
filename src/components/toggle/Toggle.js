import React from 'react';
import PropTypes from 'prop-types';

import { Container, Input, Label, Thumb } from './Toggle.styles';

const Toggle = (props) => {
  const { checked, disabled, name, id, value, onClick, ...remainingProps } = props;
  const isDisabled = disabled || !onClick;

  return (
    <Container data-testid="container" checked={checked} disabled={isDisabled}>
      <Label htmlFor={id} disabled={isDisabled}>
        <Input
          data-testid="toggle"
          type="checkbox"
          id={id}
          name={name}
          value={value}
          checked={checked}
          disabled={isDisabled}
          onClick={isDisabled ? undefined : (ev) => onClick(ev.target)}
          {...remainingProps}
        />
      </Label>
      <Thumb />
    </Container>
  );
};

Toggle.propTypes = {
  /** Determine if the toggle is on or off */
  checked: PropTypes.bool.isRequired,
  /** Disables the toggle */
  disabled: PropTypes.bool,
  /** Name of the toggle that is used on forms */
  name: PropTypes.string,
  /** Id of the toggle that is used on forms */
  id: PropTypes.string,
  /** Value of toggle */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.bool, PropTypes.number]),
  /** Callback to be fired when toggle is pressed */
  onClick: PropTypes.func,
};

Toggle.defaultProps = {
  checked: false,
  disabled: false,
};

export default Toggle;
