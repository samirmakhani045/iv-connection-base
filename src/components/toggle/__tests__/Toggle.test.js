import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Toggle from '../Toggle';

const defaultProps = {
  checked: true,
  disabled: false,
};

const Test = (props) => <Toggle {...defaultProps} {...props} />;

describe('<Toggle />', () => {
  it('should render correctly', () => {
    const { container } = render(<Test />);

    expect(screen.queryByTestId('toggle')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a disabled toggle', () => {
    const { container } = render(<Test disabled />);

    expect(screen.queryByTestId('toggle')).toBeDisabled();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a checked toggle', () => {
    const { container } = render(<Test />);

    expect(screen.queryByTestId('toggle')).toBeChecked();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a disabled and checked toggle', () => {
    const { container } = render(<Test disabled />);

    expect(screen.queryByTestId('toggle')).toBeDisabled();
    expect(screen.queryByTestId('toggle')).toBeChecked();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should fire onClick if toggle has callback', () => {
    const onClick = jest.fn();
    render(<Test onClick={onClick} />);

    const toggle = screen.queryByTestId('toggle');

    userEvent.click(toggle);
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('should not fire onClick if toggle is disabled', () => {
    const onClick = jest.fn();
    render(<Test onClick={onClick} disabled />);

    const toggle = screen.queryByTestId('toggle');

    userEvent.click(toggle);
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});
