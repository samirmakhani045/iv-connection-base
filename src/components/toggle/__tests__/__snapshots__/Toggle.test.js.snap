// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render a checked toggle 1`] = `
<div
  class="Togglestyles__Container-ku0wh7-0 jxlaHN"
  data-testid="container"
  disabled=""
>
  <label
    class="Togglestyles__Label-ku0wh7-2 hCTgVf"
    disabled=""
  >
    <input
      checked=""
      class="Togglestyles__Input-ku0wh7-1 gsgTNP"
      data-testid="toggle"
      disabled=""
      type="checkbox"
      value=""
    />
  </label>
  <div
    class="Togglestyles__Thumb-ku0wh7-3 tYcUG"
  />
</div>
`;

exports[`<Toggle /> should render a disabled and checked toggle 1`] = `
<div
  class="Togglestyles__Container-ku0wh7-0 jxlaHN"
  data-testid="container"
  disabled=""
>
  <label
    class="Togglestyles__Label-ku0wh7-2 hCTgVf"
    disabled=""
  >
    <input
      checked=""
      class="Togglestyles__Input-ku0wh7-1 gsgTNP"
      data-testid="toggle"
      disabled=""
      type="checkbox"
      value=""
    />
  </label>
  <div
    class="Togglestyles__Thumb-ku0wh7-3 tYcUG"
  />
</div>
`;

exports[`<Toggle /> should render a disabled toggle 1`] = `
<div
  class="Togglestyles__Container-ku0wh7-0 jxlaHN"
  data-testid="container"
  disabled=""
>
  <label
    class="Togglestyles__Label-ku0wh7-2 hCTgVf"
    disabled=""
  >
    <input
      checked=""
      class="Togglestyles__Input-ku0wh7-1 gsgTNP"
      data-testid="toggle"
      disabled=""
      type="checkbox"
      value=""
    />
  </label>
  <div
    class="Togglestyles__Thumb-ku0wh7-3 tYcUG"
  />
</div>
`;

exports[`<Toggle /> should render correctly 1`] = `
<div
  class="Togglestyles__Container-ku0wh7-0 jxlaHN"
  data-testid="container"
  disabled=""
>
  <label
    class="Togglestyles__Label-ku0wh7-2 hCTgVf"
    disabled=""
  >
    <input
      checked=""
      class="Togglestyles__Input-ku0wh7-1 gsgTNP"
      data-testid="toggle"
      disabled=""
      type="checkbox"
      value=""
    />
  </label>
  <div
    class="Togglestyles__Thumb-ku0wh7-3 tYcUG"
  />
</div>
`;
